rfcs/0003-storage.md

```diff
@@ -1,46 +1,107 @@
+- (d) replicator can subscribe to an abbreviated transaction stream to generate the key state itself
+2. A replicator obtains the PoH hash corresponding to the last key rotation along with its entry\_height.
+3. The replicator signs the PoH hash with its keypair. That signature is the seed used to pick the block to replicate and also the encryption key. The replicator mods the signature with the entry\_height to get which block to replicate.
+4. The replicator retrives the ledger by asking peer validators and replicators. See 6.5.
+5. The replicator then encrypts that block with the key with chacha algorithm in CBC mode with NUM\_CHACHA\_ROUNDS of encryption.
+6. The replicator initializes a chacha rng with the signature from step 2 as the seed.
+7. The replicator generates NUM\_STORAGE\_SAMPLES samples in the range of the entry size and samples the encrypted block with sha256 for 32-bytes at each offset value. Sampling the state should be faster than generating the encrypted block.
+8. The replicator sends a PoRep proof transaction which contains its sha state at the end of the sampling operation, its seed and the samples it used to the current leader and it is put onto the ledger.
+9. The replicator then generates another set of offsets which it submits a fake proof with an incorrect sha state. It can be proven to be fake by providing the seed for the hash result.
+     - A fake proof should consist of a replicator hash of a signature of a PoH value. That way when the replicator reveals the fake proof, it can be verified on chain.
+10. The replicator monitors the ledger, if it sees a fake proof integrated, it creates a challenge transaction and submits it to the current leader. The transacation proves the validator incorrectly validated a fake storage proof. The replicator is rewarded and the validator's staking balance is slashed.
+
+## 7.5 Finding who has a given block of ledger
+
+1. Validators monitor the transaction stream for storage mining proofs, and keep a mapping of ledger blocks by entry\_height to public keys. When it sees a storage mining proof it updates this mapping and provides an RPC interface which takes an entry\_height and hands back a list of public keys. The client then looks up in their cluster\_info table to see which network address that corresponds to and sends a repair request to retrieve the necessary blocks of ledger.
+2. Validators would need to prune this list which it could do by periodically looking at the oldest entries in its mappings and doing a network query to see if the storage host is still serving the first entry.
+
+# 8. Sybil attacks
 
 For any random seed, we force everyone to use a signature that is derived from a PoH hash. Everyone must use the same count, so the same PoH hash is signed by every participant. The signatures are then each cryptographically tied to the keypair, which prevents a leader from grinding on the resulting value for more than 1 identity.
```

Comment:
what does "cryptographically tied to the keypair" mean?

Reply:
Anyone can verify that only someone with knowledge of the private key of the keypair created the signature from the correct message with only knowledge of the public key, signature and message.